soc/nordic/common/dmm.c

```diff
@@ -0,0 +1,272 @@
+		*buffer_out = user_buffer;
+	} else {
+		// If no:
+		// - dynamically allocate buffer in correct memory region that respects cache line
+		// alignment and padding,
+		//   using for example mem_attr_heap_alloc()
+		*buffer_out = dmm_buffer_alloc(dh, user_length);
+		// Return error if dynamic allocation fails
+		if (*buffer_out == NULL) {
+			return -ENOMEM;
+		}
+		// - copy user buffer contents into allocated buffer
+		memcpy(*buffer_out, user_buffer, user_length);
+	}
+
+	// Check if device memory region is cacheable
+	// If yes, writeback all cache lines associated with output buffer (either user or
+	// allocated)
+	if (dmm_is_cacheable(dh->region)) {
+		cache_data_flush_range(*buffer_out, user_length);
+	}
+	// If no, no action is needed
+
+	return 0;
+}
+
+int dmm_buffer_out_release(void *region, void *buffer_out)
+{
+	struct dmm_heap *dh;
+	uintptr_t addr = (uintptr_t)buffer_out;
+
+	// Get memory region that specified device can perform DMA transfers from
+	dh = dmm_heap_find((uintptr_t)region);
+	if (!dh) {
+		return -EINVAL;
+	}
+
+	// Check if output buffer is contained within memory area managed by dynamic memory
+	// allocator
+	if (dmm_is_contained_in_region(addr, 0, dmm_heap_start_get(dh), dmm_heap_size_get(dh))) {
+		// If yes, free the buffer
+		dmm_buffer_free(dh, buffer_out);
+	}
+	// If no, no action is needed
+
+	return 0;
+}
+
+int dmm_buffer_in_prepare(void *region, void *user_buffer, size_t user_length, void **buffer_in)
+{
+	struct dmm_heap *dh;
+
+	// Get memory region that specified device can perform DMA transfers to
+	dh = dmm_heap_find((uintptr_t)region);
+	if (!dh) {
+		return -EINVAL;
+	}
+
+	// Check if:
+	// - provided user buffer is already in correct memory region,
+	// - provider user buffer is aligned and padded to cache line, if it is located in cacheable
+	// region.
+	if (dmm_user_buffer_alloc_check(user_buffer, user_length, dh->region)) {
+		// If yes, assign buffer_in to user_buffer
+		*buffer_in = user_buffer;
+	} else {
+		// If no, dynamically allocate buffer in correct memory region that respects cache
+		// line alignment and padding, using for example mem_attr_heap_alloc()
+		*buffer_in = dmm_buffer_alloc(dh, user_length);
+		// Return error if dynamic allocation fails
+		if (*buffer_in == NULL) {
+			return -ENOMEM;
+		}
+	}
+
+	// Check if device memory region is cacheable
+	// If yes, invalidate all cache lines associated with output buffer (either user or
+	// allocated) to clear potential dirty bits.
+	if (dmm_is_cacheable(dh->region)) {
+		cache_data_invd_range(*buffer_in, user_length);
+	}
+	// If no, no action is needed
+
+	return 0;
+}
+
+int dmm_buffer_in_release(void *region, void *user_buffer, size_t user_length, void *buffer_in)
+{
+	struct dmm_heap *dh;
+	uintptr_t addr = (uintptr_t)buffer_in;
+
+	// Get memory region that specified device can perform DMA transfers to, using devicetree
+	dh = dmm_heap_find((uintptr_t)region);
+	if (!dh) {
+		return -EINVAL;
+	}
+
```

Comment:
I would validate the arguments here:
* `buffer_in` must be in `region`,
* if the region is cacheable `buffer_in` must be aligned

This validation can be removed in release builds for performance reasons, like with `#ifndef NDEBUG`